src/py21cmfast/drivers/coeval.py

```diff
@@ -605,44 +598,41 @@ def run_coeval(
     singleton = False
     # Ensure perturb is a list of boxes, not just one.
     if perturbed_field is None:
-        perturbed_field = []
+        perturbed_field = ()
     elif not hasattr(perturbed_field, "__len__"):
-        perturbed_field = [perturbed_field]
+        perturbed_field = (perturbed_field,)
         singleton = True
 
-    random_seed = (
-        initial_conditions.random_seed
-        if initial_conditions is not None
-        else random_seed
+    # ensure inputs are compatible with ICs/Perturbedfields
+    inputs = InputParameters.from_output_structs(
+        (initial_conditions,) + perturbed_field,
+        cosmo_params=inputs.cosmo_params,
+        user_params=inputs.user_params,
+        astro_params=inputs.astro_params,
+        flag_options=inputs.flag_options,
+        random_seed=inputs.random_seed,
     )
 
-    if isinstance(inputs, str):
-        inputs = InputParameters.from_template(inputs, seed=random_seed)
-    elif inputs is None:
-        inputs = InputParameters.from_defaults(seed=random_seed)
-    # For the high-level, we need all the InputStruct initialised
-    cosmo_params = inputs.cosmo_params
-    user_params = inputs.user_params
-    flag_options = inputs.astro_params
-    astro_params = inputs.flag_options
-
-    random_seed = inputs.random_seed
+    # if no random seed was given to the inputs, and no output structs were passed,
+    #   raise an error
+    if not isinstance(inputs.random_seed, int):
+        raise ValueError("An integer seed, or initial conditions must be given")
```

Comment:
Let's instead just make `random_seed` a required property on the `InputParameters` class.